posthog/temporal/workflows/snowflake_batch_export.py

```diff
@@ -65,23 +101,142 @@ class SnowflakeInsertInputs:
     include_events: list[str] | None = None
 
 
-def put_file_to_snowflake_table(cursor: SnowflakeCursor, file_name: str, table_name: str):
+def use_namespace(connection: SnowflakeConnection, database: str, schema: str) -> None:
+    """Switch to a namespace given by database and schema.
+
+    This allows all queries that follow to ignore database and schema.
+    """
+    cursor = connection.cursor()
+    cursor.execute(f'USE DATABASE "{database}"')
+    cursor.execute(f'USE SCHEMA "{schema}"')
+
+
+@contextlib.contextmanager
+def snowflake_connection(inputs) -> typing.Generator[SnowflakeConnection, None, None]:
+    """Context manager that yields a Snowflake connection.
+
+    Before yielding we ensure we are in the right namespace, and we set ABORT_DETACHED_QUERY
+    to FALSE to avoid Snowflake cancelling any async queries.
+    """
+    with snowflake.connector.connect(
+        user=inputs.user,
+        password=inputs.password,
+        account=inputs.account,
+        warehouse=inputs.warehouse,
+        database=inputs.database,
+        schema=inputs.schema,
+        role=inputs.role,
+    ) as connection:
+        use_namespace(connection, inputs.database, inputs.schema)
+        connection.cursor().execute("SET ABORT_DETACHED_QUERY = FALSE")
+
+        yield connection
+
+
+async def execute_async_query(
+    connection: SnowflakeConnection,
+    query: str,
+    parameters: dict | None = None,
+    file_stream=None,
+    poll_interval: float = 1.0,
+) -> str:
+    """Wrap Snowflake connector's polling API in a coroutine.
+
+    This enables asynchronous execution of queries to release the event loop to execute other tasks
+    while we poll for a query to be done. For example, the event loop may use this time for heartbeating.
+
+    Args:
+        connection: A SnowflakeConnection object as produced by snowflake.connector.connect.
+        query: A query string to run asynchronously.
+        parameters: An optional dictionary of parameters to bind to the query.
+        poll_interval: Specify how long to wait in between polls.
+    """
+    cursor = connection.cursor()
+
+    # Snowflake docs incorrectly state that the 'params' argument is named 'parameters'.
+    result = cursor.execute_async(query, params=parameters, file_stream=file_stream)
+    query_id = cursor.sfqid or result["queryId"]
+    query_status = None
+
+    try:
+        query_status = connection.get_query_status_throw_if_error(query_id)
+
+        while connection.is_still_running(query_status):
+            query_status = connection.get_query_status_throw_if_error(query_id)
+            await asyncio.sleep(poll_interval)
+
+    except snowflake.connector.ProgrammingError:
+        # TODO: logging? Other handling?
+        raise
```

Comment:
Will this exception end up in some existing log at least?

Reply:
It will bubble up and the activity will fail, so we will see it in the temporal error logs.

Reply:
The TODO was just thinking if we should do anything more, maybe log ourselves to add some context.